format.ts

```diff
@@ -1,36 +1,136 @@
-#!/usr/bin/env deno --allow-run
+#!/usr/bin/env deno --allow-run --allow-write
 // Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
+/**
+ * This script formats the source files in the repository.
+ *
+ * Usage: deno format.ts [--check]
+ *
+ * Options:
+ *   --check          Checks if the source files are formatted.
+ */
+import "./prettier/prettier.js";
+import "./prettier/parser_typescript.js";
+import "./prettier/parser_markdown.js";
+import { args, platform, readAll, exit, run, readFile, writeFile } from "deno";
+import { parse } from "./flags/mod.ts";
 
-import { readAll, exit, run } from "deno";
+// TODO: provide decent type declarions for these
+const { prettier, prettierPlugins } = window as any;
 
-async function checkVersion() {
-  const prettierVersion = run({
-    args: ["bash", "-c", "prettier --version"],
-    stdout: "piped"
+// Runs commands in cross-platform way
+function xrun(opts) {
+  return run({
+    ...opts,
+    args: platform.os === "win" ? ["cmd.exe", "/c", ...opts.args] : opts.args
   });
-  const b = await readAll(prettierVersion.stdout);
-  const s = await prettierVersion.status();
-  if (s.code != 0) {
-    console.log("error calling prettier --version error");
-    exit(s.code);
+}
+
+// Gets the source files in the repository
+async function getSourceFiles() {
+  return new TextDecoder()
+    .decode(
+      await readAll(
+        xrun({
+          args: ["git", "ls-files"],
+          stdout: "piped"
+        }).stdout
+      )
+    )
+    .trim()
+    .split(/\r?\n/);
+}
+
+/**
+ * Checks if the file has been formatted with prettier.
+ */
+async function checkFile(
+  filename: string,
+  parser: "typescript" | "markdown"
+): Promise<boolean> {
+  const text = new TextDecoder().decode(await readFile(filename));
+  const formatted = prettier.check(text, {
+    parser,
+    plugins: prettierPlugins
+  });
+
+  if (!formatted) {
+    console.log(`${filename} ... Not formatted`);
+  }
+
+  return formatted;
+}
+
+/**
+ * Formats the given file.
+ */
+async function formatFile(
+  filename: string,
+  parser: "typescript" | "markdown"
+): Promise<void> {
+  const text = new TextDecoder().decode(await readFile(filename));
+  const formatted = prettier.format(text, {
+    parser,
+    plugins: prettierPlugins
+  });
+
+  if (text !== formatted) {
+    console.log(`Formatting ${filename}`);
+    await writeFile(filename, new TextEncoder().encode(formatted));
```

Comment:
Having a file-local encoder/decoder is fine - maybe better than creating a new one each time?